File: test/language/subtypes/cpp/SubtypesTest.cpp
#include <cstdio>
#include <string>
#include <fstream>

#include "gtest/gtest.h"

#include "subtypes/TestStructure.h"
#include "subtypes/SubtypeStructure.h"
#include "subtypes/Database.h"
#include "subtypes/ParameterizedSubtypeStruct.h"

namespace subtypes
{

TEST(SubtypesTest, Uint16Subtype)
{
    const Identifier identifier = 0xFFFF;
    TestStructure testStructure;
    testStructure.setIdentifier(identifier);
    const Identifier readIdentifier = testStructure.getIdentifier();
    ASSERT_EQ(identifier, readIdentifier);
}

TEST(SubtypesTest, TestStructureSubtype)
{
    const Identifier identifier = 0xFFFF;
    const std::string name = "Name";
    Student student;
    student.setIdentifier(identifier);
    student.setName(name);

    SubtypeStructure subtypeStructure;
    subtypeStructure.setStudent(student);
    const Student readStudent = subtypeStructure.getStudent();

    ASSERT_EQ(student, readStudent);
}

TEST(SubtypesTest, TestSubtypedTable)
{
    const std::string dbFileName = "subtypes_test.sqlite";
    std::remove(dbFileName.c_str());

    Database database(dbFileName);
    database.createSchema();
    ASSERT_TRUE(database.isOpen());

    // check if database does contain the table
    std::string tableName;
    sqlite3_stmt* statement;
    const std::string sqlQuery = "SELECT name FROM sqlite_master WHERE type='table' AND name='students'";
    int result = sqlite3_prepare_v2(database.getConnection(), sqlQuery.c_str(), -1, &statement, NULL);
    ASSERT_EQ(result, SQLITE_OK);
    result = sqlite3_step(statement);
    if (result == SQLITE_ROW)
    {
        const unsigned char* readTableName = sqlite3_column_text(statement, 0);
        if (readTableName != NULL)
            tableName = reinterpret_cast<const char*>(readTableName);
    }
    sqlite3_finalize(statement);

    ASSERT_EQ("students", tableName);

    // check table getter
    TestTable& studentsAsTestTable = database.getStudents();
    SubtypedTable& studentsAsSubtypedTable = database.getStudents();
    ASSERT_EQ(&studentsAsTestTable, &studentsAsSubtypedTable);
}

namespace
{
    bool isCodeInFilePresent(const char* fileName, const char* code)
    {
        std::ifstream file(fileName);
        bool isPresent = false;
        std::string line;
        while (std::getline(file, line))
        {
            if (line.find(code) != std::string::npos)
            {
                isPresent = true;
                break;
            }
        }
        file.close();

        return isPresent;
    }
}

TEST(SubtypesTest, TestParamaterizedSubtype)
{
    // just check that ParameterizedSubtype is defined and that it's same as the ParameterizedStruct
    ParameterizedSubtypeStruct s;
    ParameterizedSubtype& parameterizedSubtype = s.getParameterizedSubtype();
    ParameterizedStruct& parameterizedStruct = s.getParameterizedSubtype();
    zserio::ObjectArray<ParameterizedSubtype>& parameterizedSubtypeArray = s.getParameterizedSubtypeArray();
    zserio::ObjectArray<ParameterizedStruct>& parameterizedStructArray = s.getParameterizedSubtypeArray();
    ASSERT_EQ(&parameterizedSubtype, &parameterizedStruct);
    ASSERT_EQ(&parameterizedSubtypeArray, &parameterizedStructArray);

    // ensure that include to the subtype is present and that the subtype is used in
    // ParameterizedSubtypeStruct's accessors
    ASSERT_TRUE(isCodeInFilePresent("language/subtypes/gen/subtypes/ParameterizedSubtypeStruct.h",
            "#include \"subtypes/ParameterizedSubtype.h\""));

    ASSERT_TRUE(isCodeInFilePresent("language/subtypes/gen/subtypes/ParameterizedSubtypeStruct.h",
            "subtypes::ParameterizedSubtype& getParameterizedSubtype()"));
    ASSERT_TRUE(isCodeInFilePresent("language/subtypes/gen/subtypes/ParameterizedSubtypeStruct.h",
            "const subtypes::ParameterizedSubtype& getParameterizedSubtype() const"));
    ASSERT_TRUE(isCodeInFilePresent("language/subtypes/gen/subtypes/ParameterizedSubtypeStruct.h",
            "void setParameterizedSubtype(const subtypes::ParameterizedSubtype& parameterizedSubtype)"));

    ASSERT_TRUE(isCodeInFilePresent("language/subtypes/gen/subtypes/ParameterizedSubtypeStruct.h",
            "zserio::ObjectArray<subtypes::ParameterizedSubtype>& getParameterizedSubtypeArray()"));
    ASSERT_TRUE(isCodeInFilePresent("language/subtypes/gen/subtypes/ParameterizedSubtypeStruct.h",
            "const zserio::ObjectArray<subtypes::ParameterizedSubtype>& getParameterizedSubtypeArray() const"));
    ASSERT_TRUE(isCodeInFilePresent("language/subtypes/gen/subtypes/ParameterizedSubtypeStruct.h",
            "void setParameterizedSubtypeArray"
                    "(const zserio::ObjectArray<subtypes::ParameterizedSubtype>& parameterizedSubtypeArray)"));
}

} // namespace subtypes

File: test/language/subtypes/java/subtypes/SubtypesTest.java
package subtypes;

import static org.junit.Assert.*;

import java.io.File;

import java.net.URISyntaxException;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;

import test_utils.FileUtil;
import test_utils.JdbcUtil;

import zserio.runtime.array.ObjectArray;

public class SubtypesTest
{
    @BeforeClass
    public static void init()
    {
        JdbcUtil.registerJdbc();
    }

    @Before
    public void setUp() throws SQLException
    {
        FileUtil.deleteFileIfExists(dbFile);
    }

    @Test
    public void uint16Subtype()
    {
        final int identifier = 0xFFFF;
        final String name = "Name";
        final TestStructure testStructure = new TestStructure(identifier, name);
        final int readIdentifier = testStructure.getIdentifier();
        assertEquals(identifier, readIdentifier);
    }

    @Test
    public void testStructureSubtype()
    {
        final int identifier = 0xFFFF;
        final String name = "Name";
        final TestStructure testStructure = new TestStructure(identifier, name);
        final SubtypeStructure subtypeStructure = new SubtypeStructure(testStructure);
        final TestStructure readTestStructure = subtypeStructure.getStudent();
        assertEquals(testStructure, readTestStructure);
    }

    @Test
    public void testSubtypedTable() throws SQLException, URISyntaxException
    {
        Database db = null;
        PreparedStatement statement = null;
        ResultSet resultSet = null;

        try
        {
            db = new Database(dbFile.toString());
            db.createSchema();
            // check if database does contain the table
            final String sqlQuery = "SELECT name FROM sqlite_master WHERE type='table' AND name='students'";
            statement = db.prepareStatement(sqlQuery);
            resultSet = statement.executeQuery();
            assertTrue(resultSet.next());
            final String tableName = resultSet.getString(1);
            assertEquals("students", tableName);

            // check table getter
            final TestTable students = db.getStudents();
            assertTrue(students != null);
        }
        catch (RuntimeException e)
        {
            fail("RuntimeException thrown: " + e.getMessage());
        }
        finally
        {
            try
            {
                if (resultSet != null)
                    resultSet.close();
            }
            finally
            {
                try
                {
                    if (statement != null)
                        statement.close();
                }
                finally
                {
                    if (db != null)
                        db.close();
                }
            }
        }
    }

    @Test
    public void testParameterizedSubtype()
    {
        // check that correct type is used instead of subtype (Java has nothing like a typedef)
        ParameterizedStruct parameterizedStruct  = new ParameterizedStruct(10);

        ParameterizedSubtypeStruct parameterizedSubtypeStruct = new ParameterizedSubtypeStruct();
        parameterizedSubtypeStruct.setLength(10);
        parameterizedSubtypeStruct.setParameterizedSubtype(parameterizedStruct);
        ObjectArray<ParameterizedStruct> parameterizedStructArray = new ObjectArray<ParameterizedStruct>(1);
        parameterizedStructArray.setElementAt(new ParameterizedStruct(10), 0);
        parameterizedSubtypeStruct.setParameterizedSubtypeArray(parameterizedStructArray);

        // it's enough that the code above compiles ok, just check something
        ParameterizedStruct parameterizedSubtype = parameterizedSubtypeStruct.getParameterizedSubtype();
        assertEquals(parameterizedSubtypeStruct.getLength(), parameterizedSubtype.getLength());
        ObjectArray<ParameterizedStruct> parameterizedSubtypeArray =
                parameterizedSubtypeStruct.getParameterizedSubtypeArray();
        assertEquals(parameterizedSubtypeStruct.getLength(),
                parameterizedSubtypeArray.elementAt(0).getLength());
    }

    private static final String DB_FILE_NAME = "subtypes_test.sqlite";

    private final File dbFile = new File(DB_FILE_NAME);
}
